// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: necode/tx/resources.proto

package tx

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort
)

// Validate checks the field values on TxId with the rules defined in the proto
// definition for this message. If any rules are violated, the first error
// encountered is returned, or nil if there are no violations.
func (m *TxId) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TxId with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in TxIdMultiError, or nil if none found.
func (m *TxId) ValidateAll() error {
	return m.validate(true)
}

func (m *TxId) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for OpaqueId

	if len(errors) > 0 {
		return TxIdMultiError(errors)
	}

	return nil
}

// TxIdMultiError is an error wrapping multiple validation errors returned by
// TxId.ValidateAll() if the designated constraints aren't met.
type TxIdMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TxIdMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TxIdMultiError) AllErrors() []error { return m }

// TxIdValidationError is the validation error returned by TxId.Validate if the
// designated constraints aren't met.
type TxIdValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TxIdValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TxIdValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TxIdValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TxIdValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TxIdValidationError) ErrorName() string { return "TxIdValidationError" }

// Error satisfies the builtin error interface
func (e TxIdValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTxId.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TxIdValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TxIdValidationError{}

// Validate checks the field values on TxInfo with the rules defined in the
// proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *TxInfo) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on TxInfo with the rules defined in the
// proto definition for this message. If any rules are violated, the result is
// a list of violation errors wrapped in TxInfoMultiError, or nil if none found.
func (m *TxInfo) ValidateAll() error {
	return m.validate(true)
}

func (m *TxInfo) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetId()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Id",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Id",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TxInfoValidationError{
				field:  "Id",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Status

	if all {
		switch v := interface{}(m.GetGrantee()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Grantee",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Grantee",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetGrantee()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TxInfoValidationError{
				field:  "Grantee",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetCreator()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Creator",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Creator",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCreator()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TxInfoValidationError{
				field:  "Creator",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetCtime()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Ctime",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "Ctime",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetCtime()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TxInfoValidationError{
				field:  "Ctime",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Description

	if all {
		switch v := interface{}(m.GetShareId()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "ShareId",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, TxInfoValidationError{
					field:  "ShareId",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetShareId()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return TxInfoValidationError{
				field:  "ShareId",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return TxInfoMultiError(errors)
	}

	return nil
}

// TxInfoMultiError is an error wrapping multiple validation errors returned by
// TxInfo.ValidateAll() if the designated constraints aren't met.
type TxInfoMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m TxInfoMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m TxInfoMultiError) AllErrors() []error { return m }

// TxInfoValidationError is the validation error returned by TxInfo.Validate if
// the designated constraints aren't met.
type TxInfoValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e TxInfoValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e TxInfoValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e TxInfoValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e TxInfoValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e TxInfoValidationError) ErrorName() string { return "TxInfoValidationError" }

// Error satisfies the builtin error interface
func (e TxInfoValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sTxInfo.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = TxInfoValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = TxInfoValidationError{}
